python_src/babymailgun/app.py

```diff
@@ -0,0 +1,227 @@
+
+
+@app.before_first_request
+def setup_app():
+    app.config["DB_HOST"] = get_env("DB_HOST")
+    try:
+        app.config["DB_PORT"] = int(get_env("DB_PORT"))
+    except ValueError:
+        raise ConfigTypeError(key="DB_PORT", key_type="int")
+
+    app.config["DB_NAME"] = get_env("DB_NAME")
+
+
+def validate_email(email_dict):
+    # these are not limits imposed by any RFC, but rather are
+    # here simply to keep things sane
+    to = email_dict["to"]
+    cc = email_dict["cc"]
+    bcc = email_dict["bcc"]
+    if len(to) + len(cc) + len(bcc) > MAX_RECIPIENTS:
+        raise TooManyRecipients()
+
+    if len(email_dict["subject"]) > MAX_SUBJECT_LENGTH:
+        raise SubjectTooLong()
+
+    if len(email_dict["body"]) > MAX_BODY_LENGTH:
+        raise BodyTooLong()
+
+    # I realize this is arbitrarily limiting, but for ease
+    # of printing on a command line I decided it was necessary/useful
+    if not SUBJECT_REGEX.match(email_dict["subject"]):
+        raise InvalidSubject()
+
+    if not EMAIL_REGEX.match(email_dict["from"]):
+        raise InvalidEmailAddress(email=email_dict["from"],
+                                  header="from")
+
+    for header in ["to", "cc", "bcc"]:
+        for email in email_dict[header]:
+            if not EMAIL_REGEX.match(email):
+                raise InvalidEmailAddress(email=email, header=header)
+
+
+def to_email_model(email_id, email_dict):
+    recipients = []
+
+    def to_recipients(recipients, recipient_type):
+        return [{"address": a,
+                 "type": recipient_type,
+                 "status": 0,
+                 "reason": ""} for a in recipients]
+
+    for receiver_type in ["to", "cc", "bcc"]:
+        recipients.extend(to_recipients(email_dict[receiver_type],
+                                        receiver_type))
+
+    return {"_id": email_id,
+            "headers": [],
+            "subject": email_dict["subject"],
+            "body": email_dict["body"],
+            "sender": email_dict["from"],
+            "recipients": recipients,
+            "created_at": datetime.datetime.now(),
+            "updated_at": datetime.datetime.fromtimestamp(0),
+            "status": "incomplete",
+            "reason": "",
+            "tries": 0,
+            "worker_id": None}
+
+
+def _get_db_client():
+    client = pymongo.MongoClient(app.config["DB_HOST"], app.config["DB_PORT"])
+    db = client[app.config["DB_NAME"]]
+    return db
+
+
+@app.route("/emails", methods=["GET"])
+def list_emails():
+    app.logger.debug("GET /emails")
+    db = _get_db_client()
+    emails = []
+    for email in db.emails.find():
+        emails.append({
+            "id": email["_id"],
+            "sender": email["sender"],
+            "status": email["status"],
+            "reason": email["reason"],
+            "created_at": email["created_at"],
+            "updated_at": email["updated_at"],
+            "tries": email["tries"]})
+    return flask.jsonify(emails)
```

Comment:
Are there any potential limitations of this API? What if the database contains billions of records?

Reply:
Totally. 

Regarding the billions of records, I'd want to introduce limit and offset to the query, along with pagination in the output response. I'd probably want to shard customers to different API endpoints (through a front-end load balancer) with different database shards behind them. No one API server or database cluster is likely to handle the load of an application managing billions of records, after all.

In terms of other limitations, the fact that there are no users/accounts implemented in this project, where I'm displaying everyones' everything to everyone else 😄 There's also no auth of any kind in front, so unregistered users could hit it and query emails.

I'm almost certainly have quotas and rate limits in place for different tiers of users to protect the API cluster itself. I would also likely stick some manner of cache plus a load balancer in front.

Beyond that, it's un-versioned, which would make iteration difficult and impossible to make improvements to the API that don't break customers somewhere. We may want to look at something like RAML or Swagger to provide customers with an API description while we're at it.

On the long-tail, totally not trying to prematurely optimize front, if there are resources that are consumed disproportionately more than other ones, I may even make discrete API endpoints for just those resources with some clever load balancing in front so that high volume calls are handled by dedicated hosts and the lower volume calls go elsewhere.